File: python/day18.py
ADD = 1
MUL = 2


def apply_op(op, a, b):
    if op == None:
        return a if b == None else b
    if op == ADD:
        return a + b
    elif op == MUL:
        return a * b


def parse_num(exp, pos):
    res = None
    while pos < len(exp) and '0' <= exp[pos] <= '9':
        d = ord(exp[pos]) - ord('0')
        res = res * 10 + d if res != None else d
        pos += 1
    return res, pos


def eval_exp(exp, pos=0, add_precedence=False):
    res = None
    op = None
    muls = []
    while pos < len(exp):
        c = exp[pos]
        if c == ')':
            pos += 1
            break
        elif c == '(':
            val, pos = eval_exp(exp, pos + 1, add_precedence)
            res = apply_op(op, res, val)
            op = None
        elif c == ' ':
            pos += 1
        elif c == '*':
            if add_precedence:
                muls.append(res)
                res = None
            else:
                op = MUL
            pos += 1
        elif c == '+':
            op = ADD
            pos += 1
        else:
            val, pos = parse_num(exp, pos)
            if val != None:
                res = apply_op(op, res, val)
                op = None
    for mul in muls:
        res *= mul
    return res, pos


def solution():
    expressions = [line.strip()
                   for line in open('../data/18.txt', 'r').readlines()]
    print(
        f'Answer 1: {sum(map(lambda x: eval_exp(x, 0, False)[0], expressions))}')
    print(
        f'Answer 2: {sum(map(lambda x: eval_exp(x, 0, True)[0], expressions))}')

File: rust/src/day18.rs
use crate::common;

#[derive(Debug)]
enum ArithOp {
  Add,
  Mul,
}

fn apply_op(op: Option<ArithOp>, a: Option<i64>, b: Option<i64>) -> Option<i64> {
  match op {
    Some(ArithOp::Add) => Some(a? + b?),
    Some(ArithOp::Mul) => Some(a? * b?),
    _ => {
      if a == None {
        b
      } else {
        a
      }
    }
  }
}

fn parse_num(exp: &str, pos: usize) -> (Option<i64>, usize) {
  let mut cpos = pos;
  let mut res = None;
  while cpos < exp.len() {
    let c = exp.as_bytes()[cpos] as char;
    if c < '0' || c > '9' {
      break;
    }
    let d = c as i64 - ('0' as i64);
    res = match res {
      Some(x) => Some(x * 10 + d),
      _ => Some(d),
    };
    cpos += 1;
  }
  (res, cpos)
}

fn eval_exp(exp: &str, pos: usize, add_precedence: bool) -> (Option<i64>, usize) {
  let mut res = None;
  let mut op = None;
  let mut cpos = pos;
  let mut muls = Vec::new();
  while cpos < exp.len() {
    match exp.as_bytes()[cpos] as char {
      ')' => {
        cpos += 1;
        break;
      }
      '(' => {
        let (val, pos1) = eval_exp(exp, cpos + 1, add_precedence);
        res = apply_op(op, res, val);
        cpos = pos1;
        op = None;
      }
      ' ' => cpos += 1,
      '*' => {
        if add_precedence {
          muls.push(res);
          res = None;
        } else {
          op = Some(ArithOp::Mul);
        }
        cpos += 1;
      }
      '+' => {
        op = Some(ArithOp::Add);
        cpos += 1;
      }
      _ => {
        let (val, pos1) = parse_num(exp, cpos);
        res = apply_op(op, res, val);
        op = None;
        cpos = pos1;
      }
    }
  }
  for mul in muls {
    res = Some(mul.unwrap() * res.unwrap());
  }
  (res, cpos)
}

pub(crate) fn solution() {
  let lines = common::read_lines(&common::data_file(18)).unwrap();
  let res1: i64 = lines
    .iter()
    .map(|line| eval_exp(line, 0, false).0.unwrap())
    .sum();
  println!("Answer 1: {}", res1);
  let res2: i64 = lines
    .iter()
    .map(|line| eval_exp(line, 0, true).0.unwrap())
    .sum();
  println!("Answer 2: {}", res2);
}
